pymc3/distributions/distribution.py

```diff
@@ -134,40 +234,165 @@ def __new__(cls, name, *args, **kwargs):
+            n_resize = observed.ndim - n_implied
+            resize = tuple(observed.shape[d] for d in range(n_resize))
+
+        if resize:
+            # A batch size was specified through `dims`, or implied by `observed`.
+            rv_out = change_rv_size(rv_var=rv_out, new_size=resize, expand=True)
+
+        if dims is not None:
+            # Now that we have a handle on the output RV, we can register named implied dimensions that
+            # were not yet known to the model, such that they can be used for size further downstream.
+            for di, dname in enumerate(dims[n_resize:]):
+                if not dname in model.dim_lengths:
+                    model.add_coord(dname, values=None, length=rv_out.shape[n_resize + di])
 
-        if "shape" in kwargs:
-            raise DeprecationWarning("The `shape` keyword is deprecated; use `size`.")
+        if testval is not None:
+            # Assigning the testval earlier causes trouble because the RV may not be created with the final shape already.
+            rv_out.tag.test_value = testval
 
-        transform = kwargs.pop("transform", UNSET)
+        rv_registered = model.register_rv(
+            rv_out, name, observed, total_size, dims=dims, transform=transform
+        )
 
-        rv_out = cls.dist(*args, rng=rng, **kwargs)
+        # Wrapping in specify_shape now does not break transforms:
+        if assert_shape is not None:
+            rv_registered = specify_shape(rv_registered, assert_shape)
 
-        return model.register_rv(rv_out, name, data, total_size, dims=dims, transform=transform)
+        return rv_registered
 
     @classmethod
-    def dist(cls, dist_params, **kwargs):
+    def dist(
+        cls,
+        dist_params,
+        *,
+        shape: Optional[Shape] = None,
+        size: Optional[Size] = None,
+        testval=None,
+        **kwargs,
+    ) -> RandomVariable:
+        """Creates a RandomVariable corresponding to the `cls` distribution.
 
-        testval = kwargs.pop("testval", None)
+        Parameters
+        ----------
+        dist_params
+        shape : tuple, optional
+            A tuple of sizes for each dimension of the new RV.
+        size : int, tuple, Variable, optional
+            A scalar or tuple for replicating the RV in addition
+            to its implied shape/dimensionality.
+        testval : optional
+            Test value to be attached to the output RV.
+            Must match its shape exactly.
 
-        rv_var = cls.rv_op(*dist_params, **kwargs)
+        Returns
+        -------
+        rv : RandomVariable
+            The created RV.
+        """
+        if "dims" in kwargs:
+            raise NotImplementedError("The use of a `.dist(dims=...)` API is not supported.")
+
+        shape, _, size = _validate_shape_dims_size(shape=shape, size=size)
+        ndim_supp = cls.rv_op.ndim_supp
+
+        if shape is not None:
+            ndim_expected = len(tuple(shape))
```

Comment:
Don't we know that `shape` is a tuple because of `_validate_shape_dims_size`?

Reply:
I will go over these conversions again. There's still some room for optimization.

Reply:
I split the conversion functions and made the type hints more precise.

After conversion `shape` can be `None`, `tuple` or `TensorVariable´. The conversion leaves `TensorVariable` shapes as they are, because conversion to tuple would create unnecessary Subtensor+MakeVector operations when the `shape` is used for RV creation later on.